x/name/keeper/keeper.go

```diff
@@ -316,3 +343,66 @@ func (k Keeper) addRecord(ctx sdk.Context, name string, addr sdk.AccAddress, res
 func (k Keeper) GetAuthority() string {
 	return k.authority
 }
+
+// DeleteInvalidAddressIndexEntries is only for the rust upgrade. It goes over all the address -> name entries and
+// deletes any that are no longer accurate.
+func (k Keeper) DeleteInvalidAddressIndexEntries(ctx sdk.Context) {
+	logger := k.Logger(ctx)
+	logger.Info("Checking address -> name index entries.")
+
+	keepCount := 0
+	var toDelete [][]byte
+
+	extractNameKey := func(key []byte) []byte {
+		// byte 1 is the type byte (0x05), it's ignored here.
+		// The 2nd byte is the length of the address that immediately follows it.
+		// The name key starts directly after the address, and is the rest of the key.
+		addrLen := int(key[1])
+		nameKeyStart := addrLen + 2
+		return key[nameKeyStart:]
+	}
+
+	store := ctx.KVStore(k.storeKey)
+	iter := sdk.KVStorePrefixIterator(store, types.AddressKeyPrefix)
+	defer func() {
+		if iter != nil {
+			iter.Close()
+		}
+	}()
+
+	for ; iter.Valid(); iter.Next() {
+		// If the key points to a non-existent name, delete it.
+		key := iter.Key()
+		nameKey := extractNameKey(key)
+		if !store.Has(nameKey) {
+			toDelete = append(toDelete, key)
+			continue
+		}
+
+		// If the index value and main value are different, delete the index.
+		indValBz := iter.Value()
+		mainValBz := store.Get(nameKey)
+		if !bytes.Equal(indValBz, mainValBz) {
+			toDelete = append(toDelete, key)
+			continue
+		}
+
+		keepCount++
+	}
+
+	iter.Close()
+	iter = nil
```

Comment:
I know we discussed this, but wanted to put a note in here too in case someone is looking back at this PR and comment.

An iterator (or really, any `Closable`) must always be closed even if there's an error or a panic. So it needs to be done using a `defer` usually created immediately after creating the iterator. In this function, though, we have stuff we want to do after using the iterator, so instead of just having `defer iter.Close()`, I put a nil check in the defer and set it to nil after manually closing it. This way, a) I'm certain `iter.Close()` will always be called exactly once, and b) the iterator is closed before the state modification that follows in this function.